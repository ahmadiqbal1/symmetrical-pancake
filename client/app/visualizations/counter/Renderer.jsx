import { isFinite, isString } from "lodash";
import React, { useState, useEffect, useMemo } from "react";
import cx from "classnames";
import Tooltip from "antd/lib/tooltip";
import resizeObserver from "@/services/resizeObserver";
import { RendererPropTypes } from "@/visualizations/prop-types";

import getCounterData from "./getCounterData";

import "./render.less";

function getCounterStyles(scale) {
  return {
    msTransform: `scale(${scale})`,
    MozTransform: `scale(${scale})`,
    WebkitTransform: `scale(${scale})`,
    transform: `scale(${scale})`,
  };
}

function getCounterScale(container) {
  const inner = container.firstChild;
  const scale = Math.min(container.offsetWidth / inner.offsetWidth, container.offsetHeight / inner.offsetHeight);
  return Number(isFinite(scale) ? scale : 1).toFixed(2); // keep only two decimal places
}

function renderTooltip(tooltip, children) {
  if (isString(tooltip) && tooltip !== "") {
    return (
      <Tooltip title={tooltip} mouseEnterDelay={0} mouseLeaveDelay={0}>
        {children}
      </Tooltip>
    );
  }
  return children;
}

export default function Renderer({ data, options, visualizationName }) {
  const [scale, setScale] = useState("1.00");
  const [container, setContainer] = useState(null);

  useEffect(() => {
    if (container) {
      const unwatch = resizeObserver(container, () => {
        setScale(getCounterScale(container));
      });
      return unwatch;
    }
  }, [container]);

  useEffect(() => {
    if (container) {
      // update scaling when options or data change (new formatting, values, etc.
      // may change inner container dimensions which will not be tracked by `resizeObserver`);
      setScale(getCounterScale(container));
    }
  }, [data, options, container]);

  const { counterLabel, showTrend, trendPositive, primaryValue, secondaryValue } = useMemo(
    () => getCounterData(data.rows, options, visualizationName),
    [data.rows, options, visualizationName]
  );

  return (
    <div
      className={cx("counter-visualization-container", {
        "trend-positive": showTrend && trendPositive,
        "trend-negative": showTrend && !trendPositive,
      })}>
      <div className="counter-visualization-content" ref={setContainer}>
        <div style={getCounterStyles(scale)}>
          {primaryValue.display !== null &&
            renderTooltip(
              primaryValue.tooltip,
              <div className={cx("counter-visualization-value", { "with-tooltip": primaryValue.tooltip !== null })}>
                {primaryValue.display}
              </div>
            )}
          {secondaryValue.display !== null &&
            renderTooltip(
              secondaryValue.tooltip,
              <div className={cx("counter-visualization-target", { "with-tooltip": secondaryValue.tooltip !== null })}>
                {secondaryValue.display}
              </div>
            )}
          {counterLabel !== null && <div className="counter-visualization-label">{counterLabel}</div>}
        </div>
      </div>
    </div>
  );
}

Renderer.propTypes = RendererPropTypes;
