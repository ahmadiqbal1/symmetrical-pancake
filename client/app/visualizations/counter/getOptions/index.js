import getOptions from "./v2";

export default getOptions;
