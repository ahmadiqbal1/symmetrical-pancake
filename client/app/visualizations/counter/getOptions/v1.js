import { extend } from "lodash";

const schemaVersion = 1;

const defaultOptions = {
  counterLabel: "",
  counterColName: "counter",
  rowNumber: 1,
  targetColName: null,
  targetRowNumber: 1,
  stringDecimal: 0,
  stringDecChar: ".",
  stringThouSep: ",",
  stringPrefix: null,
  stringSuffix: null,
  formatTargetValue: false,
  countRow: false,
};

export default function getOptions(options) {
  return extend({}, defaultOptions, options, { schemaVersion });
}
