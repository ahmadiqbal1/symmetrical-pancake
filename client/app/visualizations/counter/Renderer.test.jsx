import React from "react";
import enzyme from "enzyme";

import getOptions from "./getOptions";
import Renderer from "./Renderer";

function mount(options) {
  const data = {
    columns: [
      { name: "city", type: "string" },
      { name: "population", type: "number" },
    ],
    rows: [
      { city: "New York City", population: 18604000 },
      { city: "Shanghai", population: 24484000 },
      { city: "Tokyo", population: 38140000 },
    ],
  };

  options = getOptions(options, data);
  return enzyme.mount(<Renderer visualizationName="Test" data={data} options={options} context="query" />);
}

describe("Visualizations -> Counter -> Renderer", () => {
  test("Invalid column", () => {
    const el = mount({
      schemaVersion: 2,
      primaryValue: {
        type: "rowValue",
        column: "missing",
        rowNumber: 0,
        displayFormat: "{{ @@value_formatted }} / {{ @@value }}",
        showTooltip: false,
      },
      secondaryValue: {
        type: "unused",
        showTooltip: false,
      },
    });
    expect(el.find(".counter-visualization-container")).toMatchSnapshot();
  });

  test("Numeric Primary Value", () => {
    const el = mount({
      schemaVersion: 2,
      primaryValue: {
        type: "rowValue",
        column: "population",
        rowNumber: 1,
        displayFormat: "{{ @@value_formatted }} / {{ @@value }}",
        showTooltip: false,
      },
      secondaryValue: {
        type: "unused",
        showTooltip: false,
      },
    });
    expect(el.find(".counter-visualization-container")).toMatchSnapshot();
  });

  test("Non-numeric Primary Value", () => {
    const el = mount({
      schemaVersion: 2,
      primaryValue: {
        type: "rowValue",
        column: "city",
        rowNumber: 1,
        displayFormat: "{{ @@value_formatted }} / {{ @@value }}",
        showTooltip: false,
      },
      secondaryValue: {
        type: "unused",
        showTooltip: false,
      },
    });
    expect(el.find(".counter-visualization-container")).toMatchSnapshot();
  });

  test("Numeric Secondary Value", () => {
    const el = mount({
      schemaVersion: 2,
      primaryValue: {
        type: "rowValue",
        column: "population",
        rowNumber: 1,
        displayFormat: "{{ @@value_formatted }} / {{ @@value }}",
        showTooltip: false,
      },
      secondaryValue: {
        type: "rowValue",
        column: "population",
        rowNumber: 1,
        displayFormat: "{{ @@value_formatted }} / {{ @@value }}",
        showTooltip: false,
      },
    });
    expect(el.find(".counter-visualization-container")).toMatchSnapshot();
  });

  test("Non-numeric Secondary Value", () => {
    const el = mount({
      schemaVersion: 2,
      primaryValue: {
        type: "rowValue",
        column: "population",
        rowNumber: 1,
        displayFormat: "{{ @@value_formatted }} / {{ @@value }}",
        showTooltip: false,
      },
      secondaryValue: {
        type: "rowValue",
        column: "city",
        rowNumber: 1,
        displayFormat: "{{ @@value_formatted }} / {{ @@value }}",
        showTooltip: false,
      },
    });
    expect(el.find(".counter-visualization-container")).toMatchSnapshot();
  });

  test("Trend positive", () => {
    const el = mount({
      schemaVersion: 2,
      primaryValue: {
        type: "rowValue",
        column: "population",
        rowNumber: 3,
        displayFormat: "{{ @@value_formatted }} / {{ @@value }}",
        showTooltip: false,
      },
      secondaryValue: {
        type: "rowValue",
        column: "population",
        rowNumber: 1,
        displayFormat: "{{ @@value_formatted }} / {{ @@value }}",
        showTooltip: false,
      },
    });
    expect(el.find(".counter-visualization-container")).toMatchSnapshot();
  });

  test("Trend negative", () => {
    const el = mount({
      schemaVersion: 2,
      primaryValue: {
        type: "rowValue",
        column: "population",
        rowNumber: 1,
        displayFormat: "{{ @@value_formatted }} / {{ @@value }}",
        showTooltip: false,
      },
      secondaryValue: {
        type: "rowValue",
        column: "population",
        rowNumber: 3,
        displayFormat: "{{ @@value_formatted }} / {{ @@value }}",
        showTooltip: false,
      },
    });
    expect(el.find(".counter-visualization-container")).toMatchSnapshot();
  });

  test("With tooltips", () => {
    const el = mount({
      schemaVersion: 2,
      primaryValue: {
        type: "rowValue",
        column: "population",
        rowNumber: 1,
        displayFormat: "{{ @@value_formatted }} / {{ @@value }}",
        showTooltip: true,
      },
      secondaryValue: {
        type: "rowValue",
        column: "population",
        rowNumber: 1,
        displayFormat: "{{ @@value_formatted }} / {{ @@value }}",
        showTooltip: true,
      },
    });
    expect(el.find(".counter-visualization-container")).toMatchSnapshot();
  });
});
