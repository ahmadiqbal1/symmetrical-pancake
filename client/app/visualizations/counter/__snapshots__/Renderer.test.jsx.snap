// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Visualizations -> Counter -> Renderer Invalid column 1`] = `
<div
  className="counter-visualization-container"
>
  <div
    className="counter-visualization-content"
  >
    <div
      style={
        Object {
          "MozTransform": "scale(1.00)",
          "WebkitTransform": "scale(1.00)",
          "msTransform": "scale(1.00)",
          "transform": "scale(1.00)",
        }
      }
    />
  </div>
</div>
`;

exports[`Visualizations -> Counter -> Renderer Non-numeric Primary Value 1`] = `
<div
  className="counter-visualization-container"
>
  <div
    className="counter-visualization-content"
  >
    <div
      style={
        Object {
          "MozTransform": "scale(1.00)",
          "WebkitTransform": "scale(1.00)",
          "msTransform": "scale(1.00)",
          "transform": "scale(1.00)",
        }
      }
    >
      <div
        className="counter-visualization-value"
      >
        New York City / New York City
      </div>
      <div
        className="counter-visualization-label"
      >
        Test
      </div>
    </div>
  </div>
</div>
`;

exports[`Visualizations -> Counter -> Renderer Non-numeric Secondary Value 1`] = `
<div
  className="counter-visualization-container"
>
  <div
    className="counter-visualization-content"
  >
    <div
      style={
        Object {
          "MozTransform": "scale(1.00)",
          "WebkitTransform": "scale(1.00)",
          "msTransform": "scale(1.00)",
          "transform": "scale(1.00)",
        }
      }
    >
      <div
        className="counter-visualization-value"
      >
        18,604,000 / 18604000
      </div>
      <div
        className="counter-visualization-target"
      >
        New York City / New York City
      </div>
      <div
        className="counter-visualization-label"
      >
        Test
      </div>
    </div>
  </div>
</div>
`;

exports[`Visualizations -> Counter -> Renderer Numeric Primary Value 1`] = `
<div
  className="counter-visualization-container"
>
  <div
    className="counter-visualization-content"
  >
    <div
      style={
        Object {
          "MozTransform": "scale(1.00)",
          "WebkitTransform": "scale(1.00)",
          "msTransform": "scale(1.00)",
          "transform": "scale(1.00)",
        }
      }
    >
      <div
        className="counter-visualization-value"
      >
        18,604,000 / 18604000
      </div>
      <div
        className="counter-visualization-label"
      >
        Test
      </div>
    </div>
  </div>
</div>
`;

exports[`Visualizations -> Counter -> Renderer Numeric Secondary Value 1`] = `
<div
  className="counter-visualization-container trend-positive"
>
  <div
    className="counter-visualization-content"
  >
    <div
      style={
        Object {
          "MozTransform": "scale(1.00)",
          "WebkitTransform": "scale(1.00)",
          "msTransform": "scale(1.00)",
          "transform": "scale(1.00)",
        }
      }
    >
      <div
        className="counter-visualization-value"
      >
        18,604,000 / 18604000
      </div>
      <div
        className="counter-visualization-target"
      >
        18,604,000 / 18604000
      </div>
      <div
        className="counter-visualization-label"
      >
        Test
      </div>
    </div>
  </div>
</div>
`;

exports[`Visualizations -> Counter -> Renderer Trend negative 1`] = `
<div
  className="counter-visualization-container trend-negative"
>
  <div
    className="counter-visualization-content"
  >
    <div
      style={
        Object {
          "MozTransform": "scale(1.00)",
          "WebkitTransform": "scale(1.00)",
          "msTransform": "scale(1.00)",
          "transform": "scale(1.00)",
        }
      }
    >
      <div
        className="counter-visualization-value"
      >
        18,604,000 / 18604000
      </div>
      <div
        className="counter-visualization-target"
      >
        38,140,000 / 38140000
      </div>
      <div
        className="counter-visualization-label"
      >
        Test
      </div>
    </div>
  </div>
</div>
`;

exports[`Visualizations -> Counter -> Renderer Trend positive 1`] = `
<div
  className="counter-visualization-container trend-positive"
>
  <div
    className="counter-visualization-content"
  >
    <div
      style={
        Object {
          "MozTransform": "scale(1.00)",
          "WebkitTransform": "scale(1.00)",
          "msTransform": "scale(1.00)",
          "transform": "scale(1.00)",
        }
      }
    >
      <div
        className="counter-visualization-value"
      >
        38,140,000 / 38140000
      </div>
      <div
        className="counter-visualization-target"
      >
        18,604,000 / 18604000
      </div>
      <div
        className="counter-visualization-label"
      >
        Test
      </div>
    </div>
  </div>
</div>
`;

exports[`Visualizations -> Counter -> Renderer With tooltips 1`] = `
<div
  className="counter-visualization-container trend-positive"
>
  <div
    className="counter-visualization-content"
  >
    <div
      style={
        Object {
          "MozTransform": "scale(1.00)",
          "WebkitTransform": "scale(1.00)",
          "msTransform": "scale(1.00)",
          "transform": "scale(1.00)",
        }
      }
    >
      <Tooltip
        arrowPointAtCenter={false}
        autoAdjustOverflow={true}
        mouseEnterDelay={0}
        mouseLeaveDelay={0}
        placement="top"
        title="18604000"
        transitionName="zoom-big-fast"
      >
        <Tooltip
          align={Object {}}
          arrowContent={null}
          arrowPointAtCenter={false}
          autoAdjustOverflow={true}
          builtinPlacements={
            Object {
              "bottom": Object {
                "ignoreShake": true,
                "offset": Array [
                  0,
                  4,
                ],
                "overflow": Object {
                  "adjustX": 1,
                  "adjustY": 1,
                },
                "points": Array [
                  "tc",
                  "bc",
                ],
                "targetOffset": Array [
                  0,
                  0,
                ],
              },
              "bottomLeft": Object {
                "ignoreShake": true,
                "offset": Array [
                  0,
                  4,
                ],
                "overflow": Object {
                  "adjustX": 1,
                  "adjustY": 1,
                },
                "points": Array [
                  "tl",
                  "bl",
                ],
                "targetOffset": Array [
                  0,
                  0,
                ],
              },
              "bottomRight": Object {
                "ignoreShake": true,
                "offset": Array [
                  0,
                  4,
                ],
                "overflow": Object {
                  "adjustX": 1,
                  "adjustY": 1,
                },
                "points": Array [
                  "tr",
                  "br",
                ],
                "targetOffset": Array [
                  0,
                  0,
                ],
              },
              "left": Object {
                "ignoreShake": true,
                "offset": Array [
                  -4,
                  0,
                ],
                "overflow": Object {
                  "adjustX": 1,
                  "adjustY": 1,
                },
                "points": Array [
                  "cr",
                  "cl",
                ],
                "targetOffset": Array [
                  0,
                  0,
                ],
              },
              "leftBottom": Object {
                "ignoreShake": true,
                "offset": Array [
                  -4,
                  0,
                ],
                "overflow": Object {
                  "adjustX": 1,
                  "adjustY": 1,
                },
                "points": Array [
                  "br",
                  "bl",
                ],
                "targetOffset": Array [
                  0,
                  0,
                ],
              },
              "leftTop": Object {
                "ignoreShake": true,
                "offset": Array [
                  -4,
                  0,
                ],
                "overflow": Object {
                  "adjustX": 1,
                  "adjustY": 1,
                },
                "points": Array [
                  "tr",
                  "tl",
                ],
                "targetOffset": Array [
                  0,
                  0,
                ],
              },
              "right": Object {
                "ignoreShake": true,
                "offset": Array [
                  4,
                  0,
                ],
                "overflow": Object {
                  "adjustX": 1,
                  "adjustY": 1,
                },
                "points": Array [
                  "cl",
                  "cr",
                ],
                "targetOffset": Array [
                  0,
                  0,
                ],
              },
              "rightBottom": Object {
                "ignoreShake": true,
                "offset": Array [
                  4,
                  0,
                ],
                "overflow": Object {
                  "adjustX": 1,
                  "adjustY": 1,
                },
                "points": Array [
                  "bl",
                  "br",
                ],
                "targetOffset": Array [
                  0,
                  0,
                ],
              },
              "rightTop": Object {
                "ignoreShake": true,
                "offset": Array [
                  4,
                  0,
                ],
                "overflow": Object {
                  "adjustX": 1,
                  "adjustY": 1,
                },
                "points": Array [
                  "tl",
                  "tr",
                ],
                "targetOffset": Array [
                  0,
                  0,
                ],
              },
              "top": Object {
                "ignoreShake": true,
                "offset": Array [
                  0,
                  -4,
                ],
                "overflow": Object {
                  "adjustX": 1,
                  "adjustY": 1,
                },
                "points": Array [
                  "bc",
                  "tc",
                ],
                "targetOffset": Array [
                  0,
                  0,
                ],
              },
              "topLeft": Object {
                "ignoreShake": true,
                "offset": Array [
                  0,
                  -4,
                ],
                "overflow": Object {
                  "adjustX": 1,
                  "adjustY": 1,
                },
                "points": Array [
                  "bl",
                  "tl",
                ],
                "targetOffset": Array [
                  0,
                  0,
                ],
              },
              "topRight": Object {
                "ignoreShake": true,
                "offset": Array [
                  0,
                  -4,
                ],
                "overflow": Object {
                  "adjustX": 1,
                  "adjustY": 1,
                },
                "points": Array [
                  "br",
                  "tr",
                ],
                "targetOffset": Array [
                  0,
                  0,
                ],
              },
            }
          }
          destroyTooltipOnHide={false}
          mouseEnterDelay={0}
          mouseLeaveDelay={0}
          onPopupAlign={[Function]}
          onVisibleChange={[Function]}
          overlay="18604000"
          placement="top"
          prefixCls="ant-tooltip"
          title="18604000"
          transitionName="zoom-big-fast"
          trigger={
            Array [
              "hover",
            ]
          }
          visible={false}
        >
          <Trigger
            action={
              Array [
                "hover",
              ]
            }
            afterPopupVisibleChange={[Function]}
            arrowContent={null}
            arrowPointAtCenter={false}
            autoAdjustOverflow={true}
            blurDelay={0.15}
            builtinPlacements={
              Object {
                "bottom": Object {
                  "ignoreShake": true,
                  "offset": Array [
                    0,
                    4,
                  ],
                  "overflow": Object {
                    "adjustX": 1,
                    "adjustY": 1,
                  },
                  "points": Array [
                    "tc",
                    "bc",
                  ],
                  "targetOffset": Array [
                    0,
                    0,
                  ],
                },
                "bottomLeft": Object {
                  "ignoreShake": true,
                  "offset": Array [
                    0,
                    4,
                  ],
                  "overflow": Object {
                    "adjustX": 1,
                    "adjustY": 1,
                  },
                  "points": Array [
                    "tl",
                    "bl",
                  ],
                  "targetOffset": Array [
                    0,
                    0,
                  ],
                },
                "bottomRight": Object {
                  "ignoreShake": true,
                  "offset": Array [
                    0,
                    4,
                  ],
                  "overflow": Object {
                    "adjustX": 1,
                    "adjustY": 1,
                  },
                  "points": Array [
                    "tr",
                    "br",
                  ],
                  "targetOffset": Array [
                    0,
                    0,
                  ],
                },
                "left": Object {
                  "ignoreShake": true,
                  "offset": Array [
                    -4,
                    0,
                  ],
                  "overflow": Object {
                    "adjustX": 1,
                    "adjustY": 1,
                  },
                  "points": Array [
                    "cr",
                    "cl",
                  ],
                  "targetOffset": Array [
                    0,
                    0,
                  ],
                },
                "leftBottom": Object {
                  "ignoreShake": true,
                  "offset": Array [
                    -4,
                    0,
                  ],
                  "overflow": Object {
                    "adjustX": 1,
                    "adjustY": 1,
                  },
                  "points": Array [
                    "br",
                    "bl",
                  ],
                  "targetOffset": Array [
                    0,
                    0,
                  ],
                },
                "leftTop": Object {
                  "ignoreShake": true,
                  "offset": Array [
                    -4,
                    0,
                  ],
                  "overflow": Object {
                    "adjustX": 1,
                    "adjustY": 1,
                  },
                  "points": Array [
                    "tr",
                    "tl",
                  ],
                  "targetOffset": Array [
                    0,
                    0,
                  ],
                },
                "right": Object {
                  "ignoreShake": true,
                  "offset": Array [
                    4,
                    0,
                  ],
                  "overflow": Object {
                    "adjustX": 1,
                    "adjustY": 1,
                  },
                  "points": Array [
                    "cl",
                    "cr",
                  ],
                  "targetOffset": Array [
                    0,
                    0,
                  ],
                },
                "rightBottom": Object {
                  "ignoreShake": true,
                  "offset": Array [
                    4,
                    0,
                  ],
                  "overflow": Object {
                    "adjustX": 1,
                    "adjustY": 1,
                  },
                  "points": Array [
                    "bl",
                    "br",
                  ],
                  "targetOffset": Array [
                    0,
                    0,
                  ],
                },
                "rightTop": Object {
                  "ignoreShake": true,
                  "offset": Array [
                    4,
                    0,
                  ],
                  "overflow": Object {
                    "adjustX": 1,
                    "adjustY": 1,
                  },
                  "points": Array [
                    "tl",
                    "tr",
                  ],
                  "targetOffset": Array [
                    0,
                    0,
                  ],
                },
                "top": Object {
                  "ignoreShake": true,
                  "offset": Array [
                    0,
                    -4,
                  ],
                  "overflow": Object {
                    "adjustX": 1,
                    "adjustY": 1,
                  },
                  "points": Array [
                    "bc",
                    "tc",
                  ],
                  "targetOffset": Array [
                    0,
                    0,
                  ],
                },
                "topLeft": Object {
                  "ignoreShake": true,
                  "offset": Array [
                    0,
                    -4,
                  ],
                  "overflow": Object {
                    "adjustX": 1,
                    "adjustY": 1,
                  },
                  "points": Array [
                    "bl",
                    "tl",
                  ],
                  "targetOffset": Array [
                    0,
                    0,
                  ],
                },
                "topRight": Object {
                  "ignoreShake": true,
                  "offset": Array [
                    0,
                    -4,
                  ],
                  "overflow": Object {
                    "adjustX": 1,
                    "adjustY": 1,
                  },
                  "points": Array [
                    "br",
                    "tr",
                  ],
                  "targetOffset": Array [
                    0,
                    0,
                  ],
                },
              }
            }
            defaultPopupVisible={false}
            destroyPopupOnHide={false}
            focusDelay={0}
            getDocument={[Function]}
            getPopupClassNameFromAlign={[Function]}
            hideAction={Array []}
            mask={false}
            maskClosable={true}
            mouseEnterDelay={0}
            mouseLeaveDelay={0}
            onPopupAlign={[Function]}
            onPopupVisibleChange={[Function]}
            overlay="18604000"
            popup={[Function]}
            popupAlign={Object {}}
            popupClassName=""
            popupPlacement="top"
            popupStyle={Object {}}
            popupTransitionName="zoom-big-fast"
            popupVisible={false}
            prefixCls="ant-tooltip"
            showAction={Array []}
            title="18604000"
            visible={false}
          >
            <div
              className="counter-visualization-value with-tooltip"
              key="trigger"
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              18,604,000 / 18604000
            </div>
          </Trigger>
        </Tooltip>
      </Tooltip>
      <Tooltip
        arrowPointAtCenter={false}
        autoAdjustOverflow={true}
        mouseEnterDelay={0}
        mouseLeaveDelay={0}
        placement="top"
        title="18604000"
        transitionName="zoom-big-fast"
      >
        <Tooltip
          align={Object {}}
          arrowContent={null}
          arrowPointAtCenter={false}
          autoAdjustOverflow={true}
          builtinPlacements={
            Object {
              "bottom": Object {
                "ignoreShake": true,
                "offset": Array [
                  0,
                  4,
                ],
                "overflow": Object {
                  "adjustX": 1,
                  "adjustY": 1,
                },
                "points": Array [
                  "tc",
                  "bc",
                ],
                "targetOffset": Array [
                  0,
                  0,
                ],
              },
              "bottomLeft": Object {
                "ignoreShake": true,
                "offset": Array [
                  0,
                  4,
                ],
                "overflow": Object {
                  "adjustX": 1,
                  "adjustY": 1,
                },
                "points": Array [
                  "tl",
                  "bl",
                ],
                "targetOffset": Array [
                  0,
                  0,
                ],
              },
              "bottomRight": Object {
                "ignoreShake": true,
                "offset": Array [
                  0,
                  4,
                ],
                "overflow": Object {
                  "adjustX": 1,
                  "adjustY": 1,
                },
                "points": Array [
                  "tr",
                  "br",
                ],
                "targetOffset": Array [
                  0,
                  0,
                ],
              },
              "left": Object {
                "ignoreShake": true,
                "offset": Array [
                  -4,
                  0,
                ],
                "overflow": Object {
                  "adjustX": 1,
                  "adjustY": 1,
                },
                "points": Array [
                  "cr",
                  "cl",
                ],
                "targetOffset": Array [
                  0,
                  0,
                ],
              },
              "leftBottom": Object {
                "ignoreShake": true,
                "offset": Array [
                  -4,
                  0,
                ],
                "overflow": Object {
                  "adjustX": 1,
                  "adjustY": 1,
                },
                "points": Array [
                  "br",
                  "bl",
                ],
                "targetOffset": Array [
                  0,
                  0,
                ],
              },
              "leftTop": Object {
                "ignoreShake": true,
                "offset": Array [
                  -4,
                  0,
                ],
                "overflow": Object {
                  "adjustX": 1,
                  "adjustY": 1,
                },
                "points": Array [
                  "tr",
                  "tl",
                ],
                "targetOffset": Array [
                  0,
                  0,
                ],
              },
              "right": Object {
                "ignoreShake": true,
                "offset": Array [
                  4,
                  0,
                ],
                "overflow": Object {
                  "adjustX": 1,
                  "adjustY": 1,
                },
                "points": Array [
                  "cl",
                  "cr",
                ],
                "targetOffset": Array [
                  0,
                  0,
                ],
              },
              "rightBottom": Object {
                "ignoreShake": true,
                "offset": Array [
                  4,
                  0,
                ],
                "overflow": Object {
                  "adjustX": 1,
                  "adjustY": 1,
                },
                "points": Array [
                  "bl",
                  "br",
                ],
                "targetOffset": Array [
                  0,
                  0,
                ],
              },
              "rightTop": Object {
                "ignoreShake": true,
                "offset": Array [
                  4,
                  0,
                ],
                "overflow": Object {
                  "adjustX": 1,
                  "adjustY": 1,
                },
                "points": Array [
                  "tl",
                  "tr",
                ],
                "targetOffset": Array [
                  0,
                  0,
                ],
              },
              "top": Object {
                "ignoreShake": true,
                "offset": Array [
                  0,
                  -4,
                ],
                "overflow": Object {
                  "adjustX": 1,
                  "adjustY": 1,
                },
                "points": Array [
                  "bc",
                  "tc",
                ],
                "targetOffset": Array [
                  0,
                  0,
                ],
              },
              "topLeft": Object {
                "ignoreShake": true,
                "offset": Array [
                  0,
                  -4,
                ],
                "overflow": Object {
                  "adjustX": 1,
                  "adjustY": 1,
                },
                "points": Array [
                  "bl",
                  "tl",
                ],
                "targetOffset": Array [
                  0,
                  0,
                ],
              },
              "topRight": Object {
                "ignoreShake": true,
                "offset": Array [
                  0,
                  -4,
                ],
                "overflow": Object {
                  "adjustX": 1,
                  "adjustY": 1,
                },
                "points": Array [
                  "br",
                  "tr",
                ],
                "targetOffset": Array [
                  0,
                  0,
                ],
              },
            }
          }
          destroyTooltipOnHide={false}
          mouseEnterDelay={0}
          mouseLeaveDelay={0}
          onPopupAlign={[Function]}
          onVisibleChange={[Function]}
          overlay="18604000"
          placement="top"
          prefixCls="ant-tooltip"
          title="18604000"
          transitionName="zoom-big-fast"
          trigger={
            Array [
              "hover",
            ]
          }
          visible={false}
        >
          <Trigger
            action={
              Array [
                "hover",
              ]
            }
            afterPopupVisibleChange={[Function]}
            arrowContent={null}
            arrowPointAtCenter={false}
            autoAdjustOverflow={true}
            blurDelay={0.15}
            builtinPlacements={
              Object {
                "bottom": Object {
                  "ignoreShake": true,
                  "offset": Array [
                    0,
                    4,
                  ],
                  "overflow": Object {
                    "adjustX": 1,
                    "adjustY": 1,
                  },
                  "points": Array [
                    "tc",
                    "bc",
                  ],
                  "targetOffset": Array [
                    0,
                    0,
                  ],
                },
                "bottomLeft": Object {
                  "ignoreShake": true,
                  "offset": Array [
                    0,
                    4,
                  ],
                  "overflow": Object {
                    "adjustX": 1,
                    "adjustY": 1,
                  },
                  "points": Array [
                    "tl",
                    "bl",
                  ],
                  "targetOffset": Array [
                    0,
                    0,
                  ],
                },
                "bottomRight": Object {
                  "ignoreShake": true,
                  "offset": Array [
                    0,
                    4,
                  ],
                  "overflow": Object {
                    "adjustX": 1,
                    "adjustY": 1,
                  },
                  "points": Array [
                    "tr",
                    "br",
                  ],
                  "targetOffset": Array [
                    0,
                    0,
                  ],
                },
                "left": Object {
                  "ignoreShake": true,
                  "offset": Array [
                    -4,
                    0,
                  ],
                  "overflow": Object {
                    "adjustX": 1,
                    "adjustY": 1,
                  },
                  "points": Array [
                    "cr",
                    "cl",
                  ],
                  "targetOffset": Array [
                    0,
                    0,
                  ],
                },
                "leftBottom": Object {
                  "ignoreShake": true,
                  "offset": Array [
                    -4,
                    0,
                  ],
                  "overflow": Object {
                    "adjustX": 1,
                    "adjustY": 1,
                  },
                  "points": Array [
                    "br",
                    "bl",
                  ],
                  "targetOffset": Array [
                    0,
                    0,
                  ],
                },
                "leftTop": Object {
                  "ignoreShake": true,
                  "offset": Array [
                    -4,
                    0,
                  ],
                  "overflow": Object {
                    "adjustX": 1,
                    "adjustY": 1,
                  },
                  "points": Array [
                    "tr",
                    "tl",
                  ],
                  "targetOffset": Array [
                    0,
                    0,
                  ],
                },
                "right": Object {
                  "ignoreShake": true,
                  "offset": Array [
                    4,
                    0,
                  ],
                  "overflow": Object {
                    "adjustX": 1,
                    "adjustY": 1,
                  },
                  "points": Array [
                    "cl",
                    "cr",
                  ],
                  "targetOffset": Array [
                    0,
                    0,
                  ],
                },
                "rightBottom": Object {
                  "ignoreShake": true,
                  "offset": Array [
                    4,
                    0,
                  ],
                  "overflow": Object {
                    "adjustX": 1,
                    "adjustY": 1,
                  },
                  "points": Array [
                    "bl",
                    "br",
                  ],
                  "targetOffset": Array [
                    0,
                    0,
                  ],
                },
                "rightTop": Object {
                  "ignoreShake": true,
                  "offset": Array [
                    4,
                    0,
                  ],
                  "overflow": Object {
                    "adjustX": 1,
                    "adjustY": 1,
                  },
                  "points": Array [
                    "tl",
                    "tr",
                  ],
                  "targetOffset": Array [
                    0,
                    0,
                  ],
                },
                "top": Object {
                  "ignoreShake": true,
                  "offset": Array [
                    0,
                    -4,
                  ],
                  "overflow": Object {
                    "adjustX": 1,
                    "adjustY": 1,
                  },
                  "points": Array [
                    "bc",
                    "tc",
                  ],
                  "targetOffset": Array [
                    0,
                    0,
                  ],
                },
                "topLeft": Object {
                  "ignoreShake": true,
                  "offset": Array [
                    0,
                    -4,
                  ],
                  "overflow": Object {
                    "adjustX": 1,
                    "adjustY": 1,
                  },
                  "points": Array [
                    "bl",
                    "tl",
                  ],
                  "targetOffset": Array [
                    0,
                    0,
                  ],
                },
                "topRight": Object {
                  "ignoreShake": true,
                  "offset": Array [
                    0,
                    -4,
                  ],
                  "overflow": Object {
                    "adjustX": 1,
                    "adjustY": 1,
                  },
                  "points": Array [
                    "br",
                    "tr",
                  ],
                  "targetOffset": Array [
                    0,
                    0,
                  ],
                },
              }
            }
            defaultPopupVisible={false}
            destroyPopupOnHide={false}
            focusDelay={0}
            getDocument={[Function]}
            getPopupClassNameFromAlign={[Function]}
            hideAction={Array []}
            mask={false}
            maskClosable={true}
            mouseEnterDelay={0}
            mouseLeaveDelay={0}
            onPopupAlign={[Function]}
            onPopupVisibleChange={[Function]}
            overlay="18604000"
            popup={[Function]}
            popupAlign={Object {}}
            popupClassName=""
            popupPlacement="top"
            popupStyle={Object {}}
            popupTransitionName="zoom-big-fast"
            popupVisible={false}
            prefixCls="ant-tooltip"
            showAction={Array []}
            title="18604000"
            visible={false}
          >
            <div
              className="counter-visualization-target with-tooltip"
              key="trigger"
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              18,604,000 / 18604000
            </div>
          </Trigger>
        </Tooltip>
      </Tooltip>
      <div
        className="counter-visualization-label"
      >
        Test
      </div>
    </div>
  </div>
</div>
`;
